// typedoc-plugin-navigation-hooks/dist/index.js
/*!
 * typedoc-plugin-navigation-hooks v1.1.2
 * Copyright © 2025-Present Hervé Perchec 
 */

import * as td from 'typedoc';

class NavigationEvent {
  constructor(items) {
    this.items = items;
  }
  /**
   * Triggered when resolving navigation.
   */
  static RESOLVE = "resolveNavigation";
  /**
   * Triggered when navigation is resolved.
   */
  static RESOLVED = "navigationResolved";
}
class NavigationItemEvent {
  constructor(item, reflection) {
    this.item = item;
    this.reflection = reflection;
  }
  /**
   * Triggered when resolving navigation item.
   */
  static RESOLVE = "resolveNavigationItem";
}
function deepParseReflection(reflection, cb) {
  if (reflection instanceof td.ContainerReflection) {
    const cbResult = cb(reflection);
    if (cbResult) return cbResult;
    if (reflection.children?.length) {
      let result;
      for (const child of reflection.children) {
        result = deepParseReflection(child, cb);
        if (result) break;
      }
      return result;
    }
  }
  return void 0;
}
function findReflectionFromPath(projectReflection, router, path) {
  return deepParseReflection(projectReflection, (reflection) => {
    if (router.getFullUrl(reflection) === path) return reflection;
  });
}
function deepParseNavigation(item, cb) {
  const cbResult = cb(item);
  if (cbResult) return cbResult;
  if (item.children?.length) {
    let result;
    for (const child of item.children) {
      result = deepParseNavigation(child, cb);
      if (result) break;
    }
    return result;
  }
  return void 0;
}
function parseNavigation(app, projectReflection, navigation) {
  app.renderer.trigger(NavigationEvent.RESOLVE, new NavigationEvent(navigation));
  for (const item of navigation) {
    parseNavigationItem(app, projectReflection, item);
  }
  app.renderer.trigger(NavigationEvent.RESOLVED, new NavigationEvent(navigation));
}
function parseNavigationItem(app, projectReflection, item) {
  const router = app.renderer.router;
  const emitEvent = (i, r) => app.renderer.trigger(NavigationItemEvent.RESOLVE, new NavigationItemEvent(i, r));
  deepParseNavigation(item, (i) => {
    if (i.path) {
      const reflection = findReflectionFromPath(projectReflection, router, i.path);
      emitEvent(i, reflection);
    } else {
      emitEvent(i, void 0);
    }
  });
}

function load(app) {
  app.renderer.preRenderAsyncJobs.push(async (output) => {
    const projectReflection = output.project;
    if (!output.navigation) {
      try {
        output.navigation = app.renderer.theme.getNavigation(projectReflection);
      } catch (err) {
        app.logger.warn(`Unable to get navigation of current theme: "${app.renderer.themeName}"`);
        output.navigation = [];
      }
    }
    const resolvedNavigation = output.navigation || [];
    parseNavigation(app, projectReflection, resolvedNavigation);
  });
}

export { NavigationEvent, NavigationItemEvent, deepParseNavigation, deepParseReflection, load as default, findReflectionFromPath, load };
