// typedoc-plugin-navigation-title/dist/index.cjs
/*!
 * typedoc-plugin-navigation-title v1.2.3
 * Copyright © 2025-Present Hervé Perchec 
 */

'use strict';

Object.defineProperties(exports, { __esModule: { value: true }, [Symbol.toStringTag]: { value: 'Module' } });

const td = require('typedoc');
const typedocPluginNavigationHooks = require('typedoc-plugin-navigation-hooks');

function _interopNamespaceDefault(e) {
  const n = Object.create(null, { [Symbol.toStringTag]: { value: 'Module' } });
  if (e) {
    for (const k in e) {
      if (k !== 'default') {
        const d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: () => e[k]
        });
      }
    }
  }
  n.default = e;
  return Object.freeze(n);
}

const td__namespace = /*#__PURE__*/_interopNamespaceDefault(td);

const CUSTOM_TAG_NAME = "navigationTitle";

function defineOptions(app) {
  app.on(td__namespace.Application.EVENT_BOOTSTRAP_END, () => {
    app.options.setValue("blockTags", [
      .../* @__PURE__ */ new Set([
        ...app.options.getValue("blockTags"),
        `@${CUSTOM_TAG_NAME}`
      ])
    ]);
  });
}

function detectCustomTag(reflection, context) {
  const comment = reflection.comment;
  return comment?.getTag(`@${CUSTOM_TAG_NAME}`);
}
function removeCustomTag(reflection, context) {
  const comment = reflection.comment;
  if (!comment) return;
  comment.removeTags(`@${CUSTOM_TAG_NAME}`);
}
function parseCustomTag(customTag, context) {
  const { logger } = context;
  const navigationTitle = customTag.content.map((part) => part.text).join("").trim();
  if (!navigationTitle.length) {
    logger.error(`Navigation title is empty. Please provide a string as title.`);
  }
  return navigationTitle;
}
function augmentReflection(reflection, navigationTitle) {
  Object.defineProperty(reflection, "_navigationTitle", {
    value: navigationTitle
  });
  return reflection;
}
function updateNavElementTitle(item, reflection) {
  if (item.text) item.text = reflection._navigationTitle;
  if (item.title) item.title = reflection._navigationTitle;
  return item;
}

function load(app) {
  defineOptions(app);
  app.converter.on(
    td__namespace.Converter.EVENT_RESOLVE,
    (context, reflection) => {
      const customTag = detectCustomTag(reflection);
      if (customTag) {
        const navigationTitle = parseCustomTag(customTag, context);
        augmentReflection(reflection.isSignature() ? reflection.parent : reflection, navigationTitle);
        removeCustomTag(reflection);
      }
    }
  );
  app.renderer.on(typedocPluginNavigationHooks.NavigationItemEvent.RESOLVE, ({ item, reflection }) => {
    if (reflection?._navigationTitle) {
      updateNavElementTitle(item, reflection);
    }
  });
}

exports.default = load;
exports.load = load;
